lib/libc/gen/directory.3
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)directory.3	8.1 (Berkeley) 6/4/93
.\" $FreeBSD: releng/11.1/lib/libc/gen/directory.3 305567 2016-09-07 21:03:11Z jilles $
.\"
.Dd August 31, 2016
.Dt DIRECTORY 3
.Os
.Sh NAME
.Nm opendir ,
.Nm fdopendir ,
.Nm readdir ,
.Nm readdir_r ,
.Nm telldir ,
.Nm seekdir ,
.Nm rewinddir ,
.Nm closedir ,
.Nm fdclosedir ,
.Nm dirfd
.Nd directory operations
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In dirent.h
.Ft DIR *
.Fn opendir "const char *filename"
.Ft DIR *
.Fn fdopendir "int fd"
.Ft struct dirent *
.Fn readdir "DIR *dirp"
.Ft int
.Fn readdir_r "DIR *dirp" "struct dirent *entry" "struct dirent **result"
.Ft long
.Fn telldir "DIR *dirp"
.Ft void
.Fn seekdir "DIR *dirp" "long loc"
.Ft void
.Fn rewinddir "DIR *dirp"
.Ft int
.Fn closedir "DIR *dirp"
.Ft int
.Fn fdclosedir "DIR *dirp"
.Ft int
.Fn dirfd "DIR *dirp"
.Sh DESCRIPTION
.Bf -symbolic
The
.Fn readdir_r
interface is deprecated
because it cannot be used correctly unless
.Brq Va NAME_MAX
is a fixed value.
.Ef
.Pp
The
.Fn opendir
function
opens the directory named by
.Fa filename ,
associates a
.Em directory stream
with it
and
returns a pointer to be used to identify the
.Em directory stream
in subsequent operations.
The pointer
.Dv NULL
is returned if
.Fa filename
cannot be accessed, or if it cannot
.Xr malloc 3
enough memory to hold the whole thing.
.Pp
The
.Fn fdopendir
function is equivalent to the
.Fn opendir
function except that the directory is specified by a file descriptor
.Fa fd
rather than by a name.
The file offset associated with the file descriptor at the time of the call
determines which entries are returned.
.Pp
Upon successful return from
.Fn fdopendir ,
the file descriptor is under the control of the system,
and if any attempt is made to close the file descriptor,
or to modify the state of the associated description other than by means
of
.Fn closedir ,
.Fn readdir ,
.Fn readdir_r ,
or
.Fn rewinddir ,
the behavior is undefined.
Upon calling
.Fn closedir
the file descriptor is closed.
The
.Dv FD_CLOEXEC
flag is set on the file descriptor by a successful call to
.Fn fdopendir .
.Pp
The
.Fn readdir
function
returns a pointer to the next directory entry.
The directory entry remains valid until the next call to
.Fn readdir
or
.Fn closedir
on the same
.Em directory stream .
The function returns
.Dv NULL
upon reaching the end of the directory or on error.
In the event of an error,
.Va errno
may be set to any of the values documented for the
.Xr getdirentries 2
system call.
.Pp
The
.Fn readdir_r
function
provides the same functionality as
.Fn readdir ,
but the caller must provide a directory
.Fa entry
buffer to store the results in.
The buffer must be large enough for a
.Vt struct dirent
with a
.Va d_name
array with
.Brq Va NAME_MAX
+ 1 elements.
If the read succeeds,
.Fa result
is pointed at the
.Fa entry ;
upon reaching the end of the directory
.Fa result
is set to
.Dv NULL .
The
.Fn readdir_r
function
returns 0 on success or an error number to indicate failure.
.Pp
The
.Fn telldir
function
returns a token representing the current location associated with the named
.Em directory stream .
Values returned by
.Fn telldir
are good only for the lifetime of the
.Dv DIR
pointer,
.Fa dirp ,
from which they are derived.
If the directory is closed and then
reopened, prior values returned by
.Fn telldir
will no longer be valid.
Values returned by
.Fn telldir
are also invalidated by a call to
.Fn rewinddir .
.Pp
The
.Fn seekdir
function
sets the position of the next
.Fn readdir
operation on the
.Em directory stream .
The new position reverts to the one associated with the
.Em directory stream
when the
.Fn telldir
operation was performed.
.Pp
The
.Fn rewinddir
function
resets the position of the named
.Em directory stream
to the beginning of the directory.
.Pp
The
.Fn closedir
function
closes the named
.Em directory stream
and frees the structure associated with the
.Fa dirp
pointer,
returning 0 on success.
On failure, \-1 is returned and the global variable
.Va errno
is set to indicate the error.
.Pp
The
.Fn fdclosedir
function is equivalent to the
.Fn closedir
function except that this function returns directory file descriptor instead of
closing it.
.Pp
The
.Fn dirfd
function
returns the integer file descriptor associated with the named
.Em directory stream ,
see
.Xr open 2 .
.Pp
Sample code which searches a directory for entry ``name'' is:
.Bd -literal -offset indent
dirp = opendir(".");
if (dirp == NULL)
	return (ERROR);
len = strlen(name);
while ((dp = readdir(dirp)) != NULL) {
	if (dp->d_namlen == len && strcmp(dp->d_name, name) == 0) {
		(void)closedir(dirp);
		return (FOUND);
	}
}
(void)closedir(dirp);
return (NOT_FOUND);
.Ed
.Sh SEE ALSO
.Xr close 2 ,
.Xr lseek 2 ,
.Xr open 2 ,
.Xr read 2 ,
.Xr dir 5
.Sh HISTORY
The
.Fn opendir ,
.Fn readdir ,
.Fn telldir ,
.Fn seekdir ,
.Fn rewinddir ,
.Fn closedir ,
and
.Fn dirfd
functions appeared in
.Bx 4.2 .
The
.Fn fdopendir
function appeared in
.Fx 8.0 .
.Fn fdclosedir
function appeared in
.Fx 10.0 .
.Sh BUGS
The behaviour of
.Fn telldir
and
.Fn seekdir
is likely to be wrong if there are parallel unlinks happening
and the directory is larger than one page.
There is code to ensure that a
.Fn seekdir
to the location given by a 
.Fn telldir
immediately before the last 
.Fn readdir
will always set the correct location to return the same value as that last
.Fn readdir
performed.
This is enough for some applications which want to "push back the last entry read" E.g. Samba. 
Seeks back to any other location,
other than the beginning of the directory,
may result in unexpected behaviour if deletes are present.
It is hoped that this situation will be resolved with changes to
.Fn getdirentries
and the VFS.


ChangeLog
2007-01-25 Christopher Olsen <cwolsen@domainatlantic.com>
	Cleaning up the ne2k driver for future network support. 
	 
	Previous version of UbixOS had LwIP functioning, I hope this will be true in this case. 



sys/fs/ubixfs/directory.c
/*****************************************************************************************
 Copyright (c) 2002-2004 The UbixOS Project
 All rights reserved.

 Redistribution and use in source and binary forms, with or without modification, are
 permitted provided that the following conditions are met:

 Redistributions of source code must retain the above copyright notice, this list of
 conditions, the following disclaimer and the list of authors.  Redistributions in binary
 form must reproduce the above copyright notice, this list of conditions, the following
 disclaimer and the list of authors in the documentation and/or other materials provided
 with the distribution. Neither the name of the UbixOS Project nor the names of its
 contributors may be used to endorse or promote products derived from this software
 without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
 EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
 OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 $Id: directory.c 79 2016-01-11 16:21:27Z reddawg $

*****************************************************************************************/

#include <ubixfs/ubixfs.h>
#include <vfs/file.h>
#include <vfs/mount.h>
#include <lib/kmalloc.h>
#include <lib/kprintf.h>
#include <string.h>

static dirList_t dirList = 0x0;

dirList_t 
ubixFSLoadDir(char *data) {
  dirList_t tmpDir = 0x0;

  tmpDir = (dirList_t)kmalloc(sizeof(struct directoryList));

  sprintf(tmpDir->dirName,"%s",data);

  if (0x0 == dirList) {
    dirList = tmpDir;
    }
  else {
    tmpDir->next = dirList;
    tmpDir->prev = 0x0;
    dirList->prev = tmpDir;
    dirList       = tmpDir;
    }

  if (!strcmp(":",data)) {
    tmpDir->dirCache = (char *)kmalloc(0x4000);
    }

  return(tmpDir);
  }

int 
addDirEntry(struct directoryEntry *dir,fileDescriptor_t *fd) {
  uInt32 i = 0x0;
  uInt32 entries = 0x0;
  struct directoryEntry *tmp = 0x0;

  tmp = (struct directoryEntry *)kmalloc(fd->size);

  readUbixFS(fd,(char *)tmp,fd->offset,fd->size);
  entries = fd->size/sizeof(struct directoryEntry);
  for (i=0;(tmp[i].attributes != 0x0) && (i < entries);i++);

  if (i == entries) {
    tmp = (struct directoryEntry *)kmalloc(0x1000);
    i = 0x0;
    }
  else {
    fd->offset = 0x0;
    }
  memcpy(&tmp[i],dir,sizeof(struct directoryEntry));
  
  if (writeUbixFS(fd,(char *)tmp,fd->offset,fd->size) == 0x0) {
    kprintf("Error Creating Directory\n");
    }
  
  return(0x0);
  }

int 
ubixFSmkDir(char *directory,fileDescriptor_t *fd) {
  int block = 0x0;

  struct directoryEntry *dir   = 0x0;
  struct directoryEntry *entry = 0x0;
  struct ubixFSInfo *fsInfo    = fd->mp->fsInfo;

  //kprintf("Creating Directory: %s",directory);
  
  block = getFreeBlocks(1,fd);
  if (block != 0x0) {
    dir   = (struct directoryEntry *)kmalloc(UBIXFS_BLOCKSIZE_BYTES);
    entry = (struct directoryEntry *)kmalloc(sizeof(struct directoryEntry));

    entry->startCluster = block;
    entry->size         = UBIXFS_BLOCKSIZE_BYTES;
    entry->attributes   = typeDirectory;
    entry->permissions  = 0xEAA;
    sprintf(entry->fileName, directory);

    //dir->attributes = typeDirectory;
    //sprintf(dir->fileName,"Test Entry");
    
    fd->mp->device->devInfo->write(fd->mp->device->devInfo->info,
                       dir,
                       fd->mp->diskLabel->partitions[fd->mp->partition].pOffset+fsInfo->blockAllocationTable[block].realSector,
                       blockSize);
    addDirEntry(entry,fd);
    kfree(dir);
    kfree(entry);
    }

  return(0x0);
  }

/***
 $Log: directory.c,v $
 Revision 1.1.1.1  2006/06/01 12:46:17  reddawg
 ubix2

 Revision 1.2  2005/10/12 00:13:37  reddawg
 Removed

 Revision 1.1.1.1  2005/09/26 17:24:41  reddawg
 no message

 Revision 1.12  2004/08/14 11:23:02  reddawg
 Changes

 Revision 1.11  2004/08/01 17:58:39  flameshadow
 chg: fixed string allocation bug in ubixfs_cacheNew()

 Revision 1.10  2004/07/23 09:10:06  reddawg
 ubixfs: cleaned up some functions played with the caching a bit
 vfs:    renamed a bunch of functions
 cleaned up a few misc bugs

 Revision 1.9  2004/07/16 20:17:29  flameshadow
 chg: broke the ufs stuff
 chg: changed vfsRegisterFS() to accept a fileSystem struct
 chg: modified calls to vfsRegisterFS() to pass fileSystem struct

 Revision 1.8  2004/06/29 03:59:47  reddawg
 Fixed a few issues with subdirectories they are working much better now

 Revision 1.7  2004/06/28 23:12:58  reddawg
 file format now container:/path/to/file

 Revision 1.6  2004/06/04 13:20:22  reddawg
 ubixFSmkDir(): played with it a bit to see if it still worked

 Revision 1.5  2004/06/04 10:19:42  reddawg
 notes: we compile again, thank g-d anyways i was about to cry

 Revision 1.4  2004/05/19 15:20:06  reddawg
 Fixed reference problems due to changes in drive subsystem

 Revision 1.3  2004/05/19 04:07:43  reddawg
 kmalloc(size,pid) no more it is no kmalloc(size); the way it should of been

 Revision 1.2  2004/04/28 02:22:55  reddawg
 This is a fiarly large commit but we are starting to use new driver model
 all around

 Revision 1.1.1.1  2004/04/15 12:07:07  reddawg
 UbixOS v1.0

 Revision 1.6  2004/04/13 16:36:34  reddawg
 Changed our copyright, it is all now under a BSD-Style license

 END
 ***/
